python_modules/libraries/dagster-airflow/dagster_airflow/dagster_job_factory.py

```diff
@@ -248,6 +337,57 @@ def make_airflow_example_dags():
     # 'search_catalog' is missing a required position argument '_'. It is fixed in airflow v2
     patch_airflow_example_dag(dag_bag)
 
-    return make_dagster_definitions_from_airflow_dag_bag(
-        dag_bag=dag_bag, use_ephemeral_airflow_db=use_ephemeral_airflow_db
-    )
+    return make_dagster_definitions_from_airflow_dag_bag(dag_bag=dag_bag)
+
+
+def make_schedules_and_jobs_from_airflow_dag_bag(
+    dag_bag,
+    connections=None,
+):
+    """Construct Dagster Schedules and Jobs corresponding to Airflow DagBag.
+
+    Args:
+        dag_bag (DagBag): Airflow DagBag Model
+        connections (List[Connection]): List of Airflow Connections to be created in the Ephemeral
+            Airflow DB, if use_emphemeral_airflow_db is False this will be ignored.
+
+    Returns:
+        - List[ScheduleDefinition]: The generated Dagster Schedules
+        - List[JobDefinition]: The generated Dagster Jobs
+    """
+    check.inst_param(dag_bag, "dag_bag", DagBag)
+    connections = check.opt_list_param(connections, "connections", of_type=Connection)
+
+    use_unique_id = contains_duplicate_task_names(dag_bag)
+
+    job_defs = []
+    schedule_defs = []
+    count = 0
+    # To enforce predictable iteration order
+    sorted_dag_ids = sorted(dag_bag.dag_ids)
+    for dag_id in sorted_dag_ids:
+        dag = dag_bag.dags.get(dag_id)
+        if not use_unique_id:
+            job_def = make_dagster_job_from_airflow_dag(
+                dag=dag,
+                tags=None,
+                connections=connections,
+            )
+        else:
+            job_def = make_dagster_job_from_airflow_dag(
+                dag=dag,
+                tags=None,
+                unique_id=count,
+                connections=connections,
+            )
+            count += 1
+        schedule_def = make_dagster_schedule_from_airflow_dag(
+            dag=dag,
+            job_def=job_def,
+        )
+        if schedule_def:
+            schedule_defs.append(schedule_def)
+        else:
+            job_defs.append(job_def)
+
+    return schedule_defs, job_defs
```

Comment:
The conditionals in here make it seem like this is bug or at a minimum problematic. Once could easily imagine a case where `job_def` is create but _not_ added. If you look carefully that is not true. but the way this is constructed makes it look scary. I'd recommend writing in a more functional, typesafe style.

`isinstance(dag.normalized_schedule_interval, str) and is_valid_cron_schedule(cron_schedule)` factored in a helper function called `dag_is_schedule` or something that returns a bool.

that _must_ be true for `make_dagster_schedule_from_airflow_dag` to successfully execute. Then you can type that as `ScheduleDefinition` rather than `Optional[ScheduleDefinition]`

```python
  if dag_is_schedule(...):
        schedule_defs.append(make_schedule(...))
   else
       jobs_defs.append(make_job(...))
```

Reply:
https://github.com/dagster-io/dagster/pull/12067/commits/16a752a5c807cfe673e36ccad60fac6deadbd3c1 does this refactor